### ISSUE TITLE
Fine-tuned model outputs noise despite decreasing training loss

### ISSUE THREAD

**Issue by KdaiP** [QwenLM/Qwen3-TTS#39]
### Description

Hi, thanks for your awesome work — the zero-shot voice cloning is very expressive and impressive.

However, there seems to be an issue with the fine-tuning code. After training for one epoch (or more), although the training loss decreases as expected, the inference output turns into pure noise. This suggests a mismatch between the fine-tuning process and the inference pipeline.

One concrete problem appears to be a missing text projection step in the fine-tuning code:
`input_text_embedding = model.talker.model.text_embedding(input_text_ids) * text_embedding_mask`
https://github.com/QwenLM/Qwen3-TTS/blob/0c6a7cbb6c8421a46332f8c2434c7825c4c855ef/finetuning/sft_12hz.py#L93

This should instead be:
`input_text_embedding = model.talker.text_projection(
    model.talker.model.text_embedding(input_text_ids)
) * text_embedding_mask`

Without this projection, fine-tuning the 0.6B model raises a size mismatch error, because the text embedding dimension and code embedding dimension are the same in the 1.7B model, but different in the 0.6B model (one is 1024, the other is 2048).

After fixing this issue, the size mismatch error is resolved, but unfortunately the fine-tuned model still produces only noise during inference.

I’d appreciate any guidance on whether there are additional steps or differences between fine-tuning and inference that might explain this behavior.

Thanks again for the great work!

### Reproduction

The model was fine-tuned on a 15-minute Mandarin speech dataset labeled by myself.

### Logs

```shell

```

### Environment Information

pytorch=2.8

### Known Issue

- [ ] The issue hasn't been already addressed in Documentation, Issues, and Discussions.

**Comment by rekuenkdr** [comment]
Effective batch size = batch size * gradient accumulation.

It's faster if you increase batch size, so my recommendation if your Vram can afford it. 

4 batch size, 8 gradient accumulation.
2 batch size, 16 gradient accumulation 
1 batch size, 32 gradient accumulation.



The loss is being adjusted too, updating the library to latest version (is independent from the nodes) and you are all set.

We already lowered lr weeks ago even before qwen published the update, based on our own findings.

**Comment by lumos675** [comment]
Guys do you know if training Persian is possible for this model? I am trying with persian dataset but The Loss remains higher than 12.

**Comment by appledev1234** [comment]
@dariox1337 Could you please share your code for fine tuning 0.6B mode? The sft_12hz.py from repo works for 1.7B model, but does not for 0.6B model. Thanks

**Comment by dariox1337** [comment]
@appledev1234 I think the fix for finetuning 0.6B was merged weeks ago. I did the same thing the first comment mentioned:
>This should instead be:
>input_text_embedding = model.talker.text_projection( model.talker.model.text_embedding(input_text_ids) ) * text_embedding_mask

And since I already deleted qwen-tts from my system, unfortunately there is nothing to share.

**Comment by univa-HARRY** [comment]
Have there been cases where the model was successfully trained using multiple reference voices instead of a single speaker?

**Comment by what-name** [comment]
**Confirming @rekuenkdr recipe works on 0.6B-Base for English.**

Data point: a clean fine-tune after a lot of failed experiments.

- Model: `Qwen3-TTS-12Hz-0.6B-Base`
- Data: 385 synthetic clips (~34 min), English, single female voice, voice-cloned from a reference through the 1.7B model, 1s trailing silence on each clip
- Script: upstream `finetuning/sft_12hz.py` with **only** the `text_projection` wrap on line 89 from the opening post. **No shift fix, no sub-codebook loop removal.** Just the dimension fix.
- Hyperparams: `--lr 1e-7 --batch_size 2 --num_epochs 2` (gradient_accumulation=4 as in the script). Single speaker at slot 3000.

**Results**
- Epoch 0: voice recognizable but flat
- Epoch 1: voice accurate, natural emotion/presence, terminates cleanly on every test text, no artifacts, no rushed speech

**What did NOT work, in case it helps others:**
- `lr 2e-5` + 20 epochs → all epochs produce noise ("aliens talking underwater")
- `lr 2e-6` + 10 epochs → epoch 0 is ok but epochs 1+ lose the EOS token and run until `max_new_tokens`
- Applying fumyou13's shift fix + `F.cross_entropy` replacement in `modeling_qwen3_tts.py` + lr 2e-5 → loss drops to ~1.0 but audio is garbage (overfitting)
- Applying the shift fix + lr 2e-6 → same pattern as unpatched script: ep0 ok, ep1+ broken

Training loss dropped only ~0.5/epoch at 1e-7 (13.4 → 12.8 over two epochs). The "loss decreasing = good training" intuition definitely does not hold here — low loss at higher LR produces noise, high loss at 1e-7 produces a clean voice. The real knob turned out to be LR, not the shift fix.

Thanks to @rekuenkdr and @dariox1337 for posting the recipe above — saved us a lot of time once we stopped chasing the shift bug.

**Comment by Shoryaveer** [comment]
Thanks @what-name for the data and compilation of approaches. I am adding another data point here:

212 minutes of mono studio recorded WAV segments (4.4s to 30s) with no volume normalization and 1 second of silence added to the end.

MODEL=0.6B-Base
BATCH_SIZE=4 
LR=1e-8 
EPOCHS=2

with minor to no loss reduction.

I would recommend setting the batch size suited for your hardware and then tune the learning rate with 2-3 epoch. Don't chase loss reduction rather check the results after tuning. The result on epoch-1 is great the only problem is lack of pause on commas and periods because I split large files based on silence present(split on >550ms of silence).

P.S. 212 minutes excluding silence.

**Comment by ittetsu07** [comment]
[@rekuenkdr](https://github.com/rekuenkdr) can we finetune in multiple speakers datasets in one language?
